okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -69,21 +74,27 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
-
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
```

Comment:
I question if that is actually what happened. I think instead the application was configuring NPN on the socket factory and it won out over our ALPN settings.

If you are concerned about the initial handshake already being completed, then sslSocket.getSession() looks to be better than sslSocket.startHandshake().